### PR TITLE
Fix CUDA DeviceNDArray slicing

### PR DESCRIPTION
**Pull request by njwhite**
The PR's tests fail on master.

### CODE REVIEW

**Comment by njwhite** [comment]
I've pushed another fix (and test!) to disallow memcpys into non-contiguous memory. Maybe this fixes should go in a 0.36 patch, as all of them just silently return incorrect data?

File: numba/cuda/cudadrv/devicearray.py [hunk 1]
```diff
@@ -146,11 +146,26 @@ def copy_to_device(self, ary, stream=0):
         if ary.size == 0:
             # Nothing to do
             return
+        sentry_contiguous(self)
         stream = self._default_stream(stream)
         if _driver.is_device_memory(ary):
+            if self.ndim > 1 and self.flags['C_CONTIGUOUS'] != ary.flags['C_CONTIGUOUS']:
```
Comment (sklam) [inline numba/cuda/cudadrv/devicearray.py]: You don't need the `self.ndim > 1`.  Strided 1D array (e.g. `np.arange(20)[::2]`) is not contiguous.

File: numba/cuda/cudadrv/devicearray.py [hunk 2]
```diff
@@ -146,11 +146,26 @@ def copy_to_device(self, ary, stream=0):
...
+                raise ValueError("Can't copy %s-contiguous array to a %s-contiguous array" % (
+                    'C' if ary.flags['C_CONTIGUOUS'] else 'F',
+                    'C' if self.flags['C_CONTIGUOUS'] else 'F',
+                ))
+
             sz = min(self.alloc_size, ary.alloc_size)
             _driver.device_to_device(self, ary, sz, stream=stream)
         else:
+            # ensure same contiguous-nous
+            if self.ndim > 1:
```
Comment (sklam) [inline numba/cuda/cudadrv/devicearray.py]: Same as above.  

An example that will break this is:

```python
a_c = np.arange(10)
d = cuda.to_device(a_c)
arr = np.arange(20)[::2]
d.copy_to_device(arr)
np.testing.assert_array_equal(d.copy_to_host(), arr)
```

File: numba/cuda/cudadrv/devicearray.py (hunk 1 above)
Comment (sklam) [reply numba/cuda/cudadrv/devicearray.py]: for `self.flags['C_CONTIGUOUS'] != ary.flags['C_CONTIGUOUS']`, I am wondering if it is a stronger check with `self.strides != ary.strides`.

File: numba/cuda/cudadrv/devicearray.py [hunk 3]
```diff
@@ -285,12 +300,28 @@ def is_f_contiguous(self):
         '''
         return self._dummy.is_f_contig
 
+    @property
+    def flags(self):
```
Comment (sklam) [inline numba/cuda/cudadrv/devicearray.py]: You may need to add this for the `DeviceRecord` class.  There are tests failing due to missing `flags` for it.

Comment (njwhite) [reply numba/cuda/cudadrv/devicearray.py]: Which tests? The github CI builds passed?

File: numba/cuda/cudadrv/devicearray.py (hunk 2 above)
Comment (njwhite) [reply numba/cuda/cudadrv/devicearray.py]: I've added this example to the test cases

**Comment by njwhite** [comment]
@sklam I've pushed changes that (I think!) address all your review comments

File: numba/cuda/cudadrv/devicearray.py (hunk 3 above)
Comment (sklam) [reply numba/cuda/cudadrv/devicearray.py]: The TravisCI builds cannot test GPU code.  So, the GPU tests are skipped on CI.  To run all the GPU tests, use `python runtests.py -v numba.cuda.tests`.  

The failing tests were from `numba.cuda.tests.cudapy.test_record_dtype`.  They are passing now.

**Comment by sklam** [comment]
There is a new test failure:

```

======================================================================
ERROR: test_vectorize_scalar_arg (numba.cuda.tests.cudapy.test_vectorize_scalar_arg.TestCUDAVectorizeScalarArg)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/siu/dev/numba/numba/cuda/tests/cudapy/test_vectorize_scalar_arg.py", line 27, in test_vectorize_scalar_arg
    vector_add(1.0, dA)
  File "/Users/siu/dev/numba/numba/cuda/dispatcher.py", line 88, in __call__
    return CUDAUFuncMechanism.call(self.functions, args, kws)
  File "/Users/siu/dev/numba/numba/npyufunc/deviceufunc.py", line 281, in call
    dev_a = cr.to_device(a, stream=stream)
  File "/Users/siu/dev/numba/numba/cuda/dispatcher.py", line 203, in to_device
    return cuda.to_device(hostary, stream=stream)
  File "/Users/siu/dev/numba/numba/cuda/cudadrv/devices.py", line 212, in _require_cuda_context
    return fn(*args, **kws)
  File "/Users/siu/dev/numba/numba/cuda/api.py", line 58, in to_device
    to, new = devicearray.auto_device(obj, stream=stream, copy=copy)
  File "/Users/siu/dev/numba/numba/cuda/cudadrv/devicearray.py", line 511, in auto_device
    devobj.copy_to_device(obj, stream=stream)
  File "/Users/siu/dev/numba/numba/cuda/cudadrv/devices.py", line 212, in _require_cuda_context
    return fn(*args, **kws)
  File "/Users/siu/dev/numba/numba/cuda/cudadrv/devicearray.py", line 150, in copy_to_device
    sentry_contiguous(self)
  File "/Users/siu/dev/numba/numba/cuda/cudadrv/devicearray.py", line 490, in sentry_contiguous
    return sentry_contiguous(ary[0])
  File "/Users/siu/dev/numba/numba/cuda/cudadrv/devices.py", line 212, in _require_cuda_context
    return fn(*args, **kws)
  File "/Users/siu/dev/numba/numba/cuda/cudadrv/devicearray.py", line 382, in __getitem__
    return self._do_getitem(item)
  File "/Users/siu/dev/numba/numba/cuda/cudadrv/devicearray.py", line 392, in _do_getitem
    arr = self._dummy.__getitem__(item)
  File "/Users/siu/dev/numba/numba/dummyarray.py", line 226, in __getitem__
    dims = [dim.__getitem__(it) for dim, it in zip(self.dims, item)]
  File "/Users/siu/dev/numba/numba/dummyarray.py", line 226, in <listcomp>
    dims = [dim.__getitem__(it) for dim, it in zip(self.dims, item)]
  File "/Users/siu/dev/numba/numba/dummyarray.py", line 81, in __getitem__
    size = (stop - start + (stride - 1)) // stride
ZeroDivisionError: integer division or modulo by zero
```

Array-scalar may have zero stride.

**Comment by njwhite** [comment]
Sorry @sklam - pushed a fix for this too (and expanded the test to check the results!)

**Comment by sklam** [comment]
There's a merge conflict.  Can you merge master into your branch and resolve the conflict?

**Comment by njwhite** [comment]
@sklam ok, done & pushed!

**Pull request merged**